core/shared/src/main/scala/fs2/internal/CompileScope.scala

```diff
@@ -199,11 +252,91 @@ private[internal] final class CompileScope[F[_]] private (val id: Token, private
       }
     }
   }
+
+  // See docs on [[Scope#interrupt]]
+  def interrupt(cause: Either[Throwable, Unit]): F[Unit] = {
+    interruptible match {
+      case None => F.raiseError(new Throwable("Scope#interrupt called for Scope that cannot be interrupted"))
+      case Some(iCtx) =>
+        val interruptRsn  = cause.left.toOption.getOrElse(Interrupted(iCtx.interruptScopeId, 0))
+        F.flatMap(F.attempt(iCtx.promise.complete(interruptRsn))) {
+          case Right(_) =>
+            F.map(iCtx.ref.modify({ case (interrupted, signalled) => (interrupted orElse Some(interruptRsn), signalled)})) { _ => ()}
+          case Left(_) =>
+            F.unit
+        }
+    }
+  }
+
+  // See docs on [[Scope#isInterrupted]]
+  def isInterrupted: F[Boolean] = {
+    interruptible match {
+      case None => F.pure(false)
+      case Some(iCtx) =>
+        F.map(iCtx.ref.get) { case (interrupted, _) => interrupted.nonEmpty }
+    }
+  }
+
+  /**
+    * If evaluates to Some(rsn) then the current step evaluation in stream shall be interrupted by
+    * given reason. Also sets the `interrupted` flag, so this is guaranteed to yield to interrupt only once.
+    *
+    * Used when interruption stream in pure steps between `uncons`
+    */
+  def shallInterrupt: F[Option[Throwable]] = {
+    interruptible match {
+      case None => F.pure(None)
+      case Some(iCtx) =>
+        F.flatMap(iCtx.ref.get) { case (interrupted, signalled) =>
+          if (signalled || interrupted.isEmpty) F.pure(None)
+          else {
+            F.map(iCtx.ref.modify { case (int, _) => (int, true)}) { c =>
+              if (c.previous._2) None
+              else c.previous._1
+            }
+          }
+        }
+    }
+  }
+
+
+  /**
+    * When the stream is evaluated, there may be `Eval` that needs to be cancelled early, when asynchronous interruption
+    * is taking the place.
+    * This allows to augment eval so whenever this scope is interupted it will return on left the reason of interruption.
+    * If the eval completes before the scope is interrupt, then this will return `A`.
+    */
+  private[internal] def interruptibleEval[A](f: F[A]): F[Either[Throwable, A]] = {
+    interruptible match {
+      case None => F.attempt(f)
+      case Some(iCtx) =>
+        F.flatMap(iCtx.ref.get) { case (_, signalled) =>
+          if (signalled) F.attempt(f)
+          else {
+            // we assume there will be no parallel evals/interpretation in the scope.
+            // the concurrent stream evaluation (like merge, join) shall have independent interruption
+            // for every asynchronous child scope of the stream, so this assumption shall be completely safe.
+            F.flatMap(iCtx.promise.cancellableGet) { case (get, cancel) =>
+              F.flatMap(fs2.async.race(get, F.attempt(f))(iCtx.effect, iCtx.ec)) {
+                case Right(result) => F.map(cancel)(_ => result)
```

Comment:
With cats syntax this becomes `cancel.as(result)`

Reply:
was trying to avoid cats.syntax in whole CompileScope. Syntax is zero cost right?

Reply:
Generally no -- some cats syntax is zero cost but anything provided via simulacrum is not. I'm fine with not using it in `CompileScope` to avoid allocations.